pkg/engine/resource/graph/util.go

```diff
@@ -0,0 +1,257 @@
+		}
+		resourceCollection = gr.WorkloadResources
+	case dependencyCategory:
+		if gr.DependencyResources == nil {
+			gr.DependencyResources = map[string]*v1.GraphResource{}
+		}
+		resourceCollection = gr.DependencyResources
+	case otherCategory:
+		if gr.OtherResources == nil {
+			gr.OtherResources = map[string]*v1.GraphResource{}
+		}
+		resourceCollection = gr.OtherResources
+	}
+
+	// Add the resource to the selected collection.
+	resourceCollection[resource.ID] = resource
+	if gr.ResourceIndex == nil {
+		gr.ResourceIndex = map[string]*v1.ResourceEntry{}
+	}
+	// Update the global resource index with the new resource and its category.
+	gr.ResourceIndex[resource.ID] = &v1.ResourceEntry{
+		Resource: resource,
+		Category: resourceCollection,
+	}
+}
+
+// FindGroupResourceByID searches for a GraphResource by its ID in the resource index.
+// If the resource is found, it returns the corresponding GraphResource. Otherwise, it returns nil.
+func FindGroupResourceByID(gr *v1.GraphResources, id string) *v1.GraphResource {
+	if entry, found := gr.ResourceIndex[id]; found {
+		return entry.Resource
+	}
+
+	return nil
+}
+
+// FindGroupResourceCollectionByID retrieves the resource collection for a specific resource ID.
+// It returns the collection (category) to which the resource belongs or nil if the resource is not found.
+func FindGroupResourceCollectionByID(gr *v1.GraphResources, id string) map[string]*v1.GraphResource {
+	if entry, found := gr.ResourceIndex[id]; found {
+		return entry.Category
+	}
+
+	return nil
+}
+
+// GenerateGraph generate a new graph from resources in the spec before apply operation is applied.
+func GenerateGraph(resources v1.Resources, gph *v1.Graph) (*v1.Graph, error) {
+	log.Infof("Adding spec resources to graph...")
+	if gph.Resources == nil {
+		gph.Resources = &v1.GraphResources{
+			WorkloadResources:   map[string]*v1.GraphResource{},
+			DependencyResources: map[string]*v1.GraphResource{},
+			OtherResources:      map[string]*v1.GraphResource{},
+			ResourceIndex:       map[string]*v1.ResourceEntry{},
+		}
+	}
+	// Get workload resources and its dependsOn resources
+	for _, res := range resources {
+		if isResourceWorkload(&res) {
+			workload := &v1.GraphResource{
+				ID:     res.ID,
+				Status: v1.ApplyFail, // status initialized to failure before actual apply operation
+			}
+			addGraphResource(gph.Resources, workload, workloadCategory)
+
+			if len(res.DependsOn) != 0 {
+				for _, dependOn := range res.DependsOn {
+					dependOn := &v1.GraphResource{
+						ID:     dependOn,
+						Status: v1.ApplyFail, // status initialized to failure before actual apply operation
+					}
+					addGraphResource(gph.Resources, dependOn, dependencyCategory)
+				}
+			}
+			log.Infof("Added workload resource %s to graph", workload.ID)
+			break
+		}
+	}
+
+	// Put other resources to graph
+	for _, res := range resources {
+		if found := FindGroupResourceByID(gph.Resources, res.ID); found == nil {
```

Comment:
Will the `Workload` resource be processed repeatedly here?

Reply:
No, the `Workload` resource has been recorded in ResourceIndex already here, thus it will be caught by `FindGraphResource` (typo, will fix) here.